File: solutions/024-swap-nodes-in-pairs/swap-nodes-in-pairs.py
# -*- coding:utf-8 -*-


# Given a linked list, swap every two adjacent nodes and return its head.
#
# You may not modify the values in the list's nodes, only nodes itself may be changed.
#
#  
#
# Example:
#
#
# Given 1->2->3->4, you should return the list as 2->1->4->3.
#
#


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = h = ListNode(0)
        step = 1
        tmp = None
        while head:
            if step % 2:
                tmp = ListNode(head.val)
            else:
                h.next = ListNode(head.val)
                h.next.next = tmp
                h = h.next.next
                tmp = None
            head = head.next
            step += 1
        if tmp:
            h.next = tmp
        return dummy.next

File: solutions/024-swap-nodes-in-pairs/swap-nodes-in-pairs.cpp
// Given a linked list, swap every two adjacent nodes and return its head.
//
// You may not modify the values in the list's nodes, only nodes itself may be changed.
//
//  
//
// Example:
//
//
// Given 1->2->3->4, you should return the list as 2->1->4->3.
//
//


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode dummy(0);
        ListNode* fast = &dummy;
        ListNode* prev = &dummy;
        fast->next = head;
        int cnt = 0;
        while (fast->next) {
            cnt++;
            fast = fast->next;
            if (cnt == 2) {
                cnt = 0;
                ListNode* slow = prev->next;
                ListNode* todo = fast->next;
                ListNode* h = todo;
                while (slow != todo) {
                    ListNode* next = slow->next;
                    slow->next = h;
                    h = slow;
                    slow = next;
                }
                ListNode* next_p = prev->next;
                prev->next = h;
                prev = next_p;
                fast = next_p;
            }
        }
        return dummy.next;
    }
};
